lib/fluent/supervisor.rb

```diff
@@ -35,21 +37,187 @@
+        $log.debug "fluentd RPC got /api/config.dump request"
+        $log.info "dump in-memory config"
+        supervisor_dump_config_handler
+        nil
+      }
+
+      rpc_server.mount_proc('/api/config.getDump') { |req, res|
+        $log.debug "fluentd RPC got /api/config.dump request"
+        $log.info "get dump in-memory config via HTTP"
+        res.body = supervisor_get_dump_config_handler
+        [nil, nil, res]
+      } if config[:enable_get_dump]
+
+      rpc_server.start
+
+      config[:rpc_server] = rpc_server
+    end
+
+    def install_supervisor_signal_handlers
+      trap :USR1 do
+        $log.debug "fluentd supervisor process get SIGUSR1"
+        supervisor_sigusr1_handler
+      end unless Fluent.windows?
+    end
+
+    def supervisor_sigusr1_handler
+      $log.reopen!
+      if pid = config[:worker_pid]
+        Process.kill(:USR1, pid)
+        # don't rescue Erro::ESRSH here (invalid status)
+      end
+    end
+
+    def kill_worker
+      if pid = config[:worker_pid]
+        Process.kill :INT, pid
+      end
+    end
+  end
+
+  module WorkerModule
+    def spawn(process_manager)
+      main_cmd = config[:main_cmd]
+      @pm = process_manager.spawn(main_cmd)
+    end
+
+    def after_start
+      config[:worker_pid] = @pm.pid
+    end
+  end
+
+  class Supervisor
+    def self.load_config(path, params = {})
+
+      pre_loadtime = 0
+      pre_loadtime = params['pre_loadtime'].to_i if params['pre_loadtime']
+      pre_config_mtime = nil
+      pre_config_mtime = params['pre_config_mtime'] if params['pre_config_mtime']
+      config_mtime = File.mtime(path)
+
+      # reuse previous config if last load time is within 5 seconds and mtime of the config file is not changed
+      if Time.now - Time.at(pre_loadtime) < 5 and config_mtime == pre_config_mtime
+        return params['pre_conf']
+      end
+
+      config_fname = File.basename(path)
+      config_basedir = File.dirname(path)
+      config_data = File.read(path)
+      inline_config = params['inline_config']
+      if inline_config == '-'
+        config_data << "\n" << STDIN.read
+      elsif inline_config
+        config_data << "\n" << inline_config.gsub("\\n","\n")
       end
+      fluentd_conf = Fluent::Config.parse(config_data, config_fname, config_basedir, params['use_v1_config'])
+      system_config = SystemConfig.create(fluentd_conf)
+
+      log_level = system_config.log_level || params['log_level']
+      suppress_repeated_stacktrace = system_config.suppress_repeated_stacktrace || params['suppress_repeated_stacktrace']
+      log_path = params['log_path']
+      chuser = params['chuser']
+      chgroup = params['chgroup']
+      rpc_endpoint = system_config.rpc_endpoint
+      enable_get_dump = system_config.enable_get_dump
+
+      log_opts = {suppress_repeated_stacktrace: suppress_repeated_stacktrace}
+      log = Supervisor::LoggerInitializer.new(log_path, log_level, chuser, chgroup, log_opts)
```

Comment:
Finally I thought that variable name `log` is bad name. There're many `log` used as logger (and almost same about `$log`), but this `log` is not logger, but initializer.
So this name should be another one to show "this is initializer, not logger itself", and it's better to have comment to show that `.init` sets initialized logger to `$log`.